EVENTS.md

```diff
@@ -0,0 +1,416 @@
+Here are some examples:
+
+```go
+// Store Code
+sdk.NewEvent(
+    "store_code",
+    // Update in 0.18: _code_id is also a reserved prefix
+    sdk.NewAttribute("_code_id", fmt.Sprintf("%d", codeID)),
+    // features required by the contract (new in 0.18)
+    // see https://github.com/CosmWasm/wasmd/issues/574
+    sdk.NewAttribute("feature", "stargate"),
+    sdk.NewAttribute("feature", "staking"),
+)
+
+// Instantiate Contract
+sdk.NewEvent(
+    "instantiate",
+    sdk.NewAttribute("_code_id", fmt.Sprintf("%d", msg.CodeID)),
+    sdk.NewAttribute("_contract_addr", contractAddr.String()),
+    sdk.NewAttribute("result", hex.EncodeToString(data)),
+)
+
+// Execute Contract
+sdk.NewEvent(
+    "execute",
+    sdk.NewAttribute("_contract_addr", contractAddr.String()),
+    sdk.NewAttribute("result", hex.EncodeToString(data)),
+)
+
+// Migrate Contract
+sdk.NewEvent(
+    "migrate",
+    // Note: this is the new code id that is being migrated to
+    sdk.NewAttribute("_code_id", fmt.Sprintf("%d", msg.CodeID)),
+    sdk.NewAttribute("_contract_addr", contractAddr.String()),
+    sdk.NewAttribute("result", hex.EncodeToString(data)),
+)
+
+// Set new admin
+sdk.NewEvent(
+    "update_admin",
+    sdk.NewAttribute("_contract_addr", contractAddr.String()),
+    sdk.NewAttribute("admin", msg.NewAdmin),
+)
+
+// Clear admin
+sdk.NewEvent(
+    "clear_admin",
+    sdk.NewAttribute("_contract_addr", contractAddr.String()),
+)
+
+// Pin Code
+sdk.NewEvent(
+    "pin_code",
+    sdk.NewAttribute("_code_id", strconv.FormatUint(msg.CodeID, 10)),
+)
+
+// Unpin Code
+sdk.NewEvent(
+    "unpin_code",
+    sdk.NewAttribute("_code_id", strconv.FormatUint(msg.CodeID, 10)),
+)
+
+// Emitted when processing a submessage reply
+sdk.NewEvent(
+    "reply",
+    sdk.NewAttribute("_contract_addr", contractAddr.String()),
+)
+
+```
+
+Note that every event that affects a contract (not store code, pin or unpin) will return the contract_addr as
+`_contract_addr`. The events that are related to a particular wasm code (store code, instantiate, pin, unpin, and migrate)
+will emit that as `_code_id`. All attributes prefixed with `_` are reserved and may not be emitted by a smart contract,
+so we use consistently with the underscore prefix, as they may also be present in the wasm events.
+
+### Emitted Custom Events from a Contract
+
+When a CosmWasm contract returns a `Response` from one of the calls, it may return a list of attributes as well as a list
+of events (in addition to data and a list of messages to dispatch). These are then processed in `x/wasm` to create events that
+are emitted to the blockchain.
+
+Every contract execution, be it execute, instantiate, migrate, reply, will receive a `wasm` type event. This event will
+always be tagged with `_contract_address` by the Go module, so this is trust-worthy. The contract itself cannot overwrite
+this field. (QUESTION: do we want to emit `_code_id` as well for the code id that was just executed?) Beyond this, if the
```

Comment:
The data would be available but I would prefer to keep the attributes to a minimum.
If people are interested in historical code versions we have the migration event and also the code history.

Reply:
Fair enough, it is not really needed unless someone wants to look for all cw20 transfers.

But a better query engine could handle that as well without overloading events. I will remove